apps/hash-graph/libs/graph/src/store/postgres/knowledge/entity/mod.rs

```diff
@@ -457,275 +467,255 @@ where
+                    confidence: link_data.right_entity_confidence,
+                    provenance: link_data.right_entity_provenance.clone(),
+                });
+                link_data
+            });
+
+            entities.push(Entity {
+                properties: params.properties,
+                link_data,
+                metadata: EntityMetadata {
+                    record_id: EntityRecordId {
+                        entity_id,
+                        edition_id: entity_edition_id,
+                    },
+                    temporal_versioning,
+                    entity_type_ids: params.entity_type_ids,
+                    archived: false,
+                    provenance: entity_provenance,
+                    confidence: params.confidence,
+                    properties: params.property_metadata,
+                },
+            });
+
+            let current_num_relationships = relationships.len();
+            relationships.extend(
+                params
+                    .relationships
+                    .into_iter()
+                    .chain(once(EntityRelationAndSubject::Owner {
+                        subject: EntityOwnerSubject::Web {
+                            id: params.owned_by_id,
+                        },
+                        level: 0,
+                    }))
+                    .map(|relation_and_subject| (entity_id, relation_and_subject)),
+            );
+            if relationships.len() == current_num_relationships {
+                return Err(Report::new(InsertionError)
+                    .attach_printable("At least one relationship must be provided"));
+            }
         }
 
-        for entity_type_id in &params.entity_type_ids {
-            let entity_type_id = EntityTypeId::from_url(entity_type_id);
-            self.authorization_api
-                .check_entity_type_permission(
-                    actor_id,
-                    EntityTypePermission::Instantiate,
-                    entity_type_id,
-                    Consistency::FullyConsistent,
-                )
-                .await
-                .change_context(InsertionError)?
-                .assert_permission()
-                .change_context(InsertionError)
-                .attach(StatusCode::PermissionDenied)?;
+        let (instantiate_permissions, zookie) = self
+            .authorization_api
+            .check_entity_types_permission(
+                actor_id,
+                EntityTypePermission::Instantiate,
+                entity_type_ids,
+                Consistency::FullyConsistent,
+            )
+            .await
+            .change_context(InsertionError)?;
+        if instantiate_permissions.values().any(Not::not) {
+            return Err(Report::new(InsertionError)
+                .attach(StatusCode::PermissionDenied)
+                .attach_printable(
+                    "The actor does not have permission to instantiate one or more entity types",
```

Comment:
Will the types denied be in the context? If not it'd be helpful to add

Reply:
No, they're not. Sadly it's not straightforward to add them as we use the UUIDv5 to identify them in SpiceDB, but I can create a map on an error case. As an alternative, we could at least put the UUID in the context so it could be looked up in the database.

Reply:
Fixed in [`bbcdbe0` (#4398)](https://github.com/hashintel/hash/pull/4398/commits/bbcdbe0c5bf1fd11393bd375684d1db8d91d6fe3)